design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+* breaking change for Rook users with no clear migration path, which would be very disruptive to our current user base.
+
+#### CustomResourceDefinitions (CRDS)
+
+CRDs are what Rook is already using to extend Kubernetes.  They are a limited extension mechanism that allows the definition of custom types, but lacks the rich features of API aggregation.  For example, validation of a users CRD is only at the schema level and simple property value checks that are available via the [OpenAPI v3 schema](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject).
+Also, there is currently no versioning (conversion) support.
+However, CRDs are being actively supported by the community and more features are being added to CRDs going forward (e.g., a [versioning proposal](https://github.com/kubernetes/features/issues/544)).
+Finally, CRDs are not a breaking change for Rook users since they are already in use today.
+
+### Controller Options
+
+The controllers are the entities that will perform orchestration and deployment tasks to ensure the users desired state is made a reality within their cluster.
+There are a few options for deploying and hosting the controllers that will perform this work, as explained in the following sections.
+
+#### In-proc Controllers
+
+Multiple controllers could be deployed within a single process.
+For instance, Rook could run one controller per domain of expertise, e.g., ceph-controller, minio-controller, etc.
+Controllers would all watch the same custom resource types via a `SharedInformer` and respond to events via a `WorkQueue` for efficiency and reduced burden on core apiserver.
+Even though all controllers are watching, only the applicable controller responds to and handles an event.
+For example, the user runs `kubectl create cluster.yaml` to create a `cluster.rook.io` instance that has Ceph specific properties.
+All controllers will receive the created event via the `SharedInformer`, but only the Ceph controller will queue and handle it.
+We can consider only loading the controllers the user specifically asks for, perhaps via an environment variable in `rook-operator.yaml`.
+
+Note that this architecture can be used with either API aggregation or CRDs.
+
+##### PROS
+
+* Potentially easier developer burden for new storage backends since there is no need to create a new deployment to host their controller.
+* Less resource burden on K8s cluster since watchers/caches are being shared.
+* Easier migration path to API aggregation in the future, if CRDs usage is continued now.
+
+##### CONS
+
+* All controllers must use same base image since they are all running in the same process.
+* If a controller needs to access a backend specific tool then they will have to schedule a job that invokes the appropriate image.  This is similar to exec’ing a new process but at the cluster level.
+  * We have very limited need for this so far with the Ceph implementation (e.g., creating the initial crush map).
```

Comment:
Each storage cluster should largely manage itself. There may be some vendor independent high-level policy configuration items *that* might apply to multiple vendors, but even agreeing on those definitions would take years. If in doubt please review the efforts to standardize Unix.

With NexentaEdge what we want is to be given cluster-wide policy, which we can accept ourselves if needed, a bunch of resources (machines, NICS, RAM, drives) and be allowed to make our own decisions on how to apply those resources to providing service. From studiying lots of other storage clusters I believe that this is common.

Each solution will have specific configuration needs.  Having software written by someone other than the storage vendor properly onfigure these resources is something any of use would welcom as long as the configuration was done correctly.  But the chances of two vendors agreeing that s single configuration tool was correct for both of them is close to zero.

Reply:
For instance, NexentaEdge doesn't need to build "crash map" because it does it automatically based on distributed self discovery of currently available zones, nodes and vdevs. But I would agree that this is the area which better be handled in its own backend's container.